### PR TITLE
Fix #10520 (issue with "Save & Close" in modal) + function to update input title when changed

### LINKED ISSUE #10520: Can save Article or Category without valid name in modal view
It looks like it is possible to save an article or category without a valid name, while working in the modal view.

#### Steps to reproduce the issue
Install latest staging and multilanguage functionality, since the tab Associations is needed for this test.
Make sure you have 2 articles or 2 categories in 2 different languages.
Open an article and go to tab Associations. Make sure the Edit-button is present.

screen shot 2016-05-16 at 15 48 21

Edit the associated article via that Edit-button, clear the name of the article and click 'Save'.
Saving is not possible after this click, a red warning will be shown, as expected.

Click 'Save and Close'. Now the article will be saved with an empty name and no warning will be shown.

#### Expected result
A warning should have been showed: Invalid field: name.

#### Actual result
No warning about the missing name of the article, the article can be saved.
As a result the article is locked in the overview of articles.

#### System information (as much as possible)
Windows 10 / WAMP

#### Additional comments
This issue seems to be present in all modal views while editing a topic in such a view.

### PR DESCRIPTION
**Pull request by cyrez**
Pull Request for Issue #10520.
#### Summary of Changes
- Fix modal closing for already reviewed and refactoried edit modals.
- Article and module Select and Edit modals will be reviewed soon with the same improvements and fixes as already merged for com_categories, com_newsfeeds and com_contact, and i will add then this patch in the same time of those upcoming PRs.
- **EDIT:** add function to update input title when changed
#### Testing Instructions
- Test with edit modal : <code>newsfeed</code>, <code>contact</code> and <code>categories</code>
- Click on <code>Save</code> and <code>Save & Close</code> with a correct edition (no error), and then with an empty title, or any other possible error (not filled required field).
- Expected behavior: If an error, <code>Save & Close</code> won't close the modal, and <code>Save</code> will never close the modal (close only when requested, and on successful edition)
- More detailled testing intructions for reviewed modals here: https://github.com/joomla/joomla-cms/pull/10441
- **EDIT:** Edit a contact, news feed and/or category, change title, and save and close : this title in the input should be then updated with the new title.

### CODE REVIEW

**Comment by BurtNL** [comment]
I have tested this item :white_check_mark: successfully on 2e2af7c25e7ef43f2dbcc15143db91a323042986<br /><br />Tested successfully on categories for articles, contacts and newsfeeds. Works as described. Great!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10522">issues.joomla.org/joomla-cms/10522</a>.</sub>

**Comment by infograf768** [comment]
Works OK for all categories.
The edit button is broken for contact and newsfeed individual items: it just does not load the modal anymore here.

**Comment by JoomliC** [comment]
@infograf768 did you test on latest staging ?
As it is a fix for already mergerd changes, and contact and newsfeed were merged only yesterday (just before this issue report).

**Comment by infograf768** [comment]
Sure, on latest staging

The error here comes from these changes:
```
-				. ' href="#contactEdit' . $this->id . 'Modal"'
+				. ' href="#contactEdit' . $this->value . 'Modal"'
```
and
```
 		$html[] = JHtml::_(
 			'bootstrap.renderModal',
-			'categoryEdit' . $this->id . 'Modal',
+			'categoryEdit' . $this->value . 'Modal',
```
After changing back to $this->id the modals load OK with an href of the type
`#contactEditjform_associations_en_GBModal`

**Comment by andrepereiradasilva** [comment]
hum, i had the exact same problem @infograf768 (and used the latest staging).

@JoomliC also, i notice a small detail not related to this PR.
Just letting you now in case you didn't notice.
When i change the category name from "Photo Gallery" to "Photo Gallery test" the article gets updates, ok. But, the article title in the zone where we have the "Select" "Edit" "Clear" doesn't get updated.
image

**Comment by andrepereiradasilva** [comment]
@JoomliC the "Edit" modals work on menu items, but not in associations.

**Comment by JoomliC** [comment]
Ok i see!
I will update this PR in a few hours (have to go) but the reason is alias passed to the value, so adding (int) before $this->value will solve it.
I will do extra checking when back, to be sure ;-)

**Comment by JoomliC** [comment]
@andrepereiradasilva @BurtNL @infograf768 PR updated and should work now with associations as well as in menus ;-)

> @JoomliC also, i notice a small detail not related to this PR.
Just letting you know in case you didn't notice.
When i change the category name from "Photo Gallery" to "Photo Gallery test" the category gets updated, ok. But, the category title in the zone where we have the "Select" "Edit" "Clear" doesn't get updated.
image

@andrepereiradasilva Added too a new function for edit modal, to update the input title when changed ;-)
So to be tested too!

**Comment by andrepereiradasilva** [comment]
I have tested this item :white_check_mark: successfully on 2214bdc3a3e52847396a4eb4d6538a369ef63672<br /><br />Done tests in:
- "Select" and "Edit" buttons in Single Contact and Single News feed menu item types.
- "Select", "Edit" and "Clear" buttons in newsfeed, contact and category associations.

All worked without issues!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10522">issues.joomla.org/joomla-cms/10522</a>.</sub>

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on 2214bdc3a3e52847396a4eb4d6538a369ef63672<br /><br />Works great now.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10522">issues.joomla.org/joomla-cms/10522</a>.</sub>

**Comment by infograf768** [comment]
RTc. Thanks! Let's go on for the other modals. :smiley:<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10522">issues.joomla.org/joomla-cms/10522</a>.</sub>

File: administrator/components/com_categories/models/fields/modal/category.php [hunk 1]
```diff
@@ -48,6 +48,16 @@ protected function getInput()
...
+		// The active category id field.
+		if (0 == (int) $this->value)
+		{
+			$value = '';
+		}
+		else
+		{
+			$value = (int) $this->value;
+		}
+
```
Comment (rdeutz) [inline administrator/components/com_categories/models/fields/modal/category.php]: Why so complicated?

$value = (int) $this->value;

if (0 == $value)
{
	$value = '';
}

Comment (andrepereiradasilva) [inline administrator/components/com_categories/models/fields/modal/category.php]: Doesn't this also work?
```php
$value = ((int) $this->value === 0 ? '' : $this->value);
```

**Comment by andrepereiradasilva** [comment]
@JoomliC,

So after this PR gets merged, reggarding modals, that i see there is, the main remaining modals (for 3.6.0?):
- The latest changes also in Articles Edit/List modals
- The latest changes also in Modules Edit/List modals
- The same changes in Users List modal (if possible the "- No User -" option be in the modal buttons zone)

And the other less relevant modals (for 3.6.x or 3.7.0?):
- Use the new viewport height/width in the finder index modal, messages my settings, the multilanguage status module modal, the user notes list modal, the batch modals and other similiar modals (if other exist).

And some ideas:
- Converting the"Help" button popup up in a modal?
- Dynamic viewport height/width in the TinyMCE "+Article" "+Module" buttons?

Is this right? Or do i miss something?

File: administrator/components/com_categories/models/fields/modal/category.php (hunk 1 above)
Comment (JoomliC) [inline administrator/components/com_categories/models/fields/modal/category.php]: @rdeutz i don't have change all existed code, but i agree that as i've simplify some, i can this one too!
Do you agree if i change for this one:
`$value = (int) $this->value === 0 ? '' : (int) $this->value;`
?

**Comment by JoomliC** [comment]
@andrepereiradasilva You're right + a few minor improvement for com_template modals ;-)
First i will do priority update : article and module update (because of issues to fix)
Others will be much more cosmetic, consticency, and BS HTML structure standards

About Help, yes, good idea (for when i will have done others ;-) ) !
About TinyMCE, i don't know yet, as it's currently using TinyMCE modals (not Boostrap) which are not responsive. (not check yet, but think it could be possible to use BS modal instead of TinyMCE ones, as those 4 buttons are loaded by plugins.

**Comment by JoomliC** [comment]
@andrepereiradasilva @rdeutz @infograf768 @BurtNL I have updated the PR, by simplifying the active id $value control (as well commented by @rdeutz 👍 ) : 
- Remove if / else
- Set empty only if active id is > 0 (do not allow negative id)
- Use <code>$value</code> in title query request instead of <code>(int) $this->value > 0</code> as now, with my changes in this PR, the $value variable is set before.

Could review/test again ?
Thanks!

File: administrator/components/com_categories/models/fields/modal/category.php (hunk 1 above)
Comment (JoomliC) [inline administrator/components/com_categories/models/fields/modal/category.php]: In fact, just updated, with a simplified of active id variable ;-)

**Comment by andrepereiradasilva** [comment]
I have tested this item :white_check_mark: successfully on 29a6bf98d9ff354d40582c4b760f47e3ffe190bf<br /><br />works fine<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10522">issues.joomla.org/joomla-cms/10522</a>.</sub>

**Comment by JoomliC** [comment]
Even if this PR to fix newsfeeds, categories and contacts Select & Edit modals is not merged, i have created the PR for Article Select and Edit Modals with same fixes and updates applied here ;-)
https://github.com/joomla/joomla-cms/pull/10557

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on 29a6bf98d9ff354d40582c4b760f47e3ffe190bf<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10522">issues.joomla.org/joomla-cms/10522</a>.</sub>

**Comment by roland-d** [comment]
I have tested this item :red_circle: unsuccessfully on 29a6bf98d9ff354d40582c4b760f47e3ffe190bf<br /><br />@JoomliC I am marking this as not tested successfully because I am not sure if this is part of this PR or a separate issue but this catches your attention :)

When I go to Newsfeed (or Articles) and edit an item go to the Images tab, click on the Select for the first image. This brings up the image modal. Clicking the X in the corner works fine but the Cancel button doesn't do anything. I would expect the modal to close here.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10522">issues.joomla.org/joomla-cms/10522</a>.</sub>

**Comment by roland-d** [comment]
Removing RTC for now as there is an unsuccessful test.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10522">issues.joomla.org/joomla-cms/10522</a>.</sub>

**Comment by infograf768** [comment]
@roland-d 
I think you found another bug (which I confirm) unrelated to the purpose of this PR.

**Comment by JoomliC** [comment]
@roland-d I will check it in another PR, as not related to this PR. But well found! 👍 
And... the issue is already there in stable 3.5.1 !
So, will check this today ;-)

But you can set RTC again here ;-)

**Comment by roland-d** [comment]
@JoomliC Thank you for checking, will move this forward.

**Comment by roland-d** [comment]
Thanks everybody

**Comment by JoomliC** [comment]
@roland-d : PR done to fix "Cancel" button in image modal : https://github.com/joomla/joomla-cms/pull/10564
;-)

**Pull request merged**